src/frontc/cabs2cil.ml

```diff
@@ -4269,76 +4263,75 @@ and doExp (asconst: bool)   (* This expression is used as a constant *)
               fv.vname = "__builtin_choose_expr" || fv.vname = "__builtin_tgmath"
             | _ -> false
         in
-
+        let isBuiltinNan =
+          match f'' with
+          | Lval (Var fv, NoOffset) -> fv.vname = "__builtin_nan" || fv.vname = "__builtin_nanf" ||
+            fv.vname = "__builtin_nanl" || fv.vname = "__builtin_nans" || fv.vname = "__builtin_nansf" || fv.vname = "__builtin_nansl"
+          | _ -> false
+        in
+        if isBuiltinNan && asconst then
+          (* Replace call to builtin nan with computation yielding NaN *)
+          let onef = Const(CReal(0.0,FDouble,None)) in
+          let zerodivzero = mkCast (BinOp(Div,onef,onef,doubleType)) resType in
```

Comment:
Shouldn't `FDouble` and `doubleType` depend on the suffix of the function to be of the correct type?

Reply:
Yes, one could do that. Instead I always use double here and then convert to the correct type with a cast in `mkCast`. This cast should preserve the NaN-ness of the value.

Reply:
Ah, right, the `__builtin_nan*` functions are declared as builtins already with corresponding return types. I didn't realize those were already specified somewhere.